### PR TITLE
feat: add ldap and oidc sso

### PR DESCRIPTION
**Pull request by Meierschlumpf**
<br/>
<div align="center">
  
  <h3>Homarr</h3>
</div>

**Thank you for your contribution. Please ensure that your pull request meets the following pull request:**

- [x] Builds without warnings or errors (``pnpm buid``, autofix with ``pnpm format:fix``)
- [x] Pull request targets ``dev`` branch
- [x] Commits follow the [conventional commits guideline](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] No shorthand variable names are used (eg. ``x``, ``y``, ``i`` or any abbrevation)

### CODE REVIEW

File: apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx [hunk 1]
```diff
@@ -22,65 +24,154 @@ import { useScopedI18n } from "@homarr/translation/client";
...
-        showErrorNotification({
-          title: t("action.login.notification.error.title"),
-          message: t("action.login.notification.error.message"),
-        });
+  const credentialInputsVisible =
+    providers.includes("credentials") || providers.includes("ldap");
+
+  const onSuccess = useCallback(
+    (response: Awaited<ReturnType<typeof signIn>>) => {
+      console.log(response);
```
Comment (Meierschlumpf) [inline apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx:58]: Remove console.log

File: apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx [hunk 2]
```diff
@@ -22,65 +24,154 @@ import { useScopedI18n } from "@homarr/translation/client";
...
+      });
+    },
+    [t],
+  );
+
+  const signInAsync = useCallback(
+    async (provider: string, options?: Parameters<typeof signIn>[1]) => {
+      setIsPending(true);
+      setError(undefined);
+      console.log("signInAsync", provider, options);
```
Comment (Meierschlumpf) [inline apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx:92]: Remove console.log

File: apps/nextjs/src/app/[locale]/manage/users/[userId]/page.tsx [hunk 3]
```diff
@@ -1,5 +1,23 @@
 import { notFound } from "next/navigation";
-import { Box, Group, Stack, Title } from "@mantine/core";
+import {
```
Comment (Meierschlumpf) [inline apps/nextjs/src/app/[locale]/manage/users/[userId]/page.tsx:2]: Check imports

File: apps/nextjs/src/app/api/auth/[...nextauth]/route.ts [hunk 4]
```diff
@@ -10,5 +10,8 @@ export const POST = async (req: NextRequest) => {
 };
 
 const isCredentialsRequest = (req: NextRequest) => {
-  return req.url.includes("credentials") && req.method === "POST";
+  return (
+    (req.url.includes("credentials") || req.url.includes("ldap")) &&
```
Comment (Meierschlumpf) [inline apps/nextjs/src/app/api/auth/[...nextauth]/route.ts:14]: Ldap check  can be removed

**Comment by Meierschlumpf** [comment]
Currently the callback url stuff does not work properly for oidc as it uses the request.url for redirecting. So in docker it's redirecting to `http://df3d663300c4:3000/`. 
See https://github.com/nextauthjs/next-auth/issues/10928 for updates

File: packages/auth/env.mjs [hunk 5]
```diff
@@ -1,17 +1,114 @@
...
+  .transform((value, ctx) => {
+    const normalized = value.trim().toLowerCase();
+    if (trueStrings.includes(normalized)) return true;
+    if (falseStrings.includes(normalized)) return false;
+
+    throw new Error(`Invalid boolean value for ${ctx.path.join(".")}`);
+  });
+
+const skipValidation =
+  Boolean(process.env.CI) || Boolean(process.env.SKIP_ENV_VALIDATION);
```
[the remark this answers is not in the document]
Comment (Meierschlumpf) [reply packages/auth/env.mjs:42]: You don't want to let the ci fail when some envs are not defined. Additionally the linting for next uses the skip variable to define that it is not required for that as well.

File: packages/auth/providers/credentials/authorization/basic-authorization.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  logger.info(`user ${user.name} is trying to log in. checking password...`);
+  const isValidPassword = await bcrypt.compare(
+    credentials.password,
+    user.password,
+  );
+
+  if (!isValidPassword) {
+    logger.info(`password for user ${user.name} was incorrect`);
```
[the remark this answers is not in the document]
Comment (Meierschlumpf) [reply packages/auth/providers/credentials/authorization/basic-authorization.ts:29]: I'll change once the next-auth issue is fixed

**Comment by SeDemal** [comment]
Decision to make:
Keep ldapjs even though it's been archived, find a reputable fork, or find a new library.

**Comment by Meierschlumpf** [comment]
https://github.com/ldapts/ldapts looks promising

**Comment by manuel-rw** [comment]
@Meierschlumpf https://github.com/nextauthjs/next-auth/issues/10928 has been closed. Can you look at this again?
It's a shame that https://github.com/ldapjs/node-ldapjs had to discontinue due to toxicity.
If you had to estimate, how much effort is replacing it with https://github.com/ldapts/ldapts ?
If less than 4 hours, I would do it in this PR. If not, I would do a follow up.

**Comment by Meierschlumpf** [comment]
1. I guess I'm gonna try the provided work arround for now.
2. Regarding ldapTS, I think the library is similary build as the ldapjs library so it could just be as simple as drop and replace. I don't think it will take a lot of time. Gonna continue on this pr then this weekend

**Comment by SeDemal** [comment]
For some reason when I looked into it in the past I remembered ldapts being the typesafe version of node-ldapjs.
It may not be a direct drop and replace but it looks really close from what I can see.

**Comment by Meierschlumpf** [comment]
I was able to fix the redirect to the docker internal address by using the provided workarround from the issue mentioned above. This part now works well. Next I'm gonna make the ldap migration

**Comment by Meierschlumpf** [comment]
Thanks a lot to @SeDemal for helping me setup lldap for testing ldap locally

File: apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx [hunk 7]
```diff
@@ -14,65 +15,138 @@ import { validation } from "@homarr/validation";
...
+      showSuccessNotification({
+        title: t("action.login.notification.success.title"),
+        message: t("action.login.notification.success.message"),
+      });
+
+      // Redirect to the callback URL if the response is defined and comes from a credentials provider (ldap or credentials). oidc is redirected automatically.
+      if (response) {
         await revalidatePathActionAsync("/");
-        router.push("/");
+        void router.push(callbackUrl);
```
Comment (Meierschlumpf) [inline apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx:53]: Can be awaited

File: apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx [hunk 8]
```diff
@@ -14,65 +15,138 @@ import { validation } from "@homarr/validation";
...
+                    {t("action.login.labelWith", { provider: "LDAP" })}
+                  </SubmitButton>
+                )}
+              </Stack>
+            </form>
+            {providers.includes("oidc") && <Divider label="OIDC" labelPosition="center" />}
+          </>
+        )}
+
+        {providers.includes("oidc") && (
```
Comment (SeDemal) [inline apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx:119]: You can add a text to say they can click the button if they don't get redirected automatically.
(only when isOidcAutoLoginEnabled is true)

**Review by SeDemal — COMMENTED** [review]
Timeout and logout URL in another PR?

File: apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx (hunk 8 above)
Comment (Meierschlumpf) [reply apps/nextjs/src/app/[locale]/auth/login/_login-form.tsx:119]: I think it should be clear enough to the user that he can click that button and the text seems unnecessary to me

**Review by SeDemal — APPROVED** [review]

**Review by manuel-rw — APPROVED** [review]
I think we need to improve logging after 1.0. Otherwise looking good

**Pull request merged**